include/dr/shp/algorithms/for_each.hpp

```diff
@@ -13,84 +13,34 @@
 
 namespace shp {
 
-template <typename ExecutionPolicy, lib::distributed_contiguous_range R,
-          typename Fn>
-void for_each(ExecutionPolicy &&policy, R &&r, Fn &&fn) {
+template <lib::distributed_contiguous_range R, typename Fn>
+void for_each(const device_policy &policy, R &&r, Fn &&fn) {
+  std::span<const cl::sycl::device> devices = policy.get_devices();
```

Comment:
We haven't defined an execution policy interface, but I'd prefer to leave it a bit more generic, rather than always requiring a `std::span`.  We could use something like

```cpp
rng::forward_range auto devices = policy.get_devices();
```

to be a bit more generic.

Reply:
ok, restored `auto`